### DISCUSSION TITLE
Different Indexes gives the identical recall and similar time for creating index

### DISCUSSION THREAD

**Discussion by OscarC9912** [milvus-io/milvus#31874] [Q&A and General discussion] [answered]
Hi there: I am writing to seek for some help.

Milvus Version: v2.4.0-rc.1 (docker standalone)
Dataset: GIST Dataset (1M Vector of Dim 960)
Storage: Local in npy files
Task: Perform Approximate Nearest Neighbour Search with different Indexes
Issue: **Different** index gives **the identical recall** and **similar time for creating index**

As the dataset size is quite big, I tried to use the **do_bulk_insert** to insert the data into entities. I am pretty sure that the data is inserted successfully.  (I check the num of entities is exactly 1M).
**And the problem is: when I use different indexes, the recall is identical over different indexes and the time to create the index are also similar. (like: 0.992 for recall and 0.51 for index time)**

**However, I when DO NOT USE the **do_bulk_insert**, the index time and recall are both as desired.**

To my understanding, it is possible that the index time could be identical (same index maybe buffered in /index_files), but I really do not have any clue of why the recall is the SAME, with diff indexes

I am not sure if it has anothing to do with the do_bulk_insert.

Thanks for yr very kind help ~

**Comment by yhmo** [comment]
Different indexes and different parameters should get different recall and index time.
Typically, the bulkinsert workflow is:
1. collection.craete_index()
2. utility.do_bulk_insert()
3. utility.wait_for_index_building_complete()
4. collection.load(_refresh = True)
5. collection.search()

**Reply by OscarC9912** [reply]
Thanks for your reply, I have figured out the issue, which is I did not load / release memory well. 

But I come up with another question regarding the **index time (the time to construct the index)**:

As the time to simply execute create_index is so similar over different indexes, I am thinking **if it is true that create_index() in Milvus is just declaring but NOT constructing the index with data**, in context of using do_bulk_insert() but not insert().

The reason I re-emphasize {using do_bulk_insert() but not insert()} is that: I notice the Milvus behaviour is quite different when I use different insert methods. 
Specifically, when simply use insert(), the time to create the index could be simply calculated by setting a timer before and after create_index(). However, if I do the same for the bulk_insert, the time is always around 0.5s over different indexes.

So, if I want to evaluate the index time, is it comprehensive to measure it with the following code ?

utility.do_bulk_insert()
**TIMER1**
collection.create_index()
utility.wait_for_index_building_complete()
collection.load()
**TIMER2**

Thanks for your reply ~

**Reply by yhmo** [reply, marked as answer]
create_index() has different behavior in two use case:
1. call create_index() after insert() ---------------------------- since some data has been persisted as sealed segments, create_index() will wait until these segments are indexed. then you will see it is like "doing index work". After this method return, the sealed segments are indexed.
2. call create_index() before insert() -------------------------- no sealed segments, the create_index() only specify an index type to the vector field and quickly return.


bulkinsert tasks are executed asynchronously, the index process is a bit special to the normal insert. Always use wait_for_index_building_complete() to check the index work.

**Reply by OscarC9912** [reply]
Thank you so much for your very detailed explanation.

I still have a question for the **bulkinsert** case. 

As you mentioned, the complete procudure to construct index should be: **bulkinsert --> wait_complete --> complete**.

However, I am not sure if it normal that **different indexes** are constructed with **very similar time**. (I am pretty sure the implementation is correct as the recall exhibits expected behavious over different indexes.)

If this is expcted, may I ask what is the rationale behind it ? Is it like index is constructed with some kind of optimization ? OR could you plz kindly explain what is the meaning of **special** in **{index process is a bit special to the normal insert}** ?

Thanks!

**Reply by xiaofan-luan** [reply]
Currently, bulk insert didn't wait until index done. So it doesn't really matter about what index you use.
To wait until all data is indexed, you can create index manually and wait for the index creation done.

This will be changed soon, bulkinsert task will wait until all index done. @bigsheeper is working on it

**Reply by OscarC9912** [reply]
Thanks for your clarification. Looking forward to that feature.

But, under bulkinsert case, could you please **clarify what is the meaning** of {you can create index **manually**}, isn't it true that there is only one way of creating index on data by calling: `create_index()`, which leads to incomplete index in bulkinsert case.

OR by manually here you mean I should also use normal `insert()` with `create_index()` for the entire construction ?

Thanks for your help ~
